File: RaftWithPropBasedTests/Distributed.Consensus.Raft.Tests/TestUtil.fs
module Distributed.Consensus.Raft.Tests.TestUtil

let waitUntil<'T> (timeoutMs: int) (condition: unit -> bool) =
    let sw = System.Diagnostics.Stopwatch.StartNew()

    let rec waitUntil (condition: unit -> bool) =
        async {
            do! Async.Sleep 50

            match sw.ElapsedMilliseconds > int64 timeoutMs, condition () with
            | true, _ -> return false
            | _, true -> return true
            | _ -> return! waitUntil condition
        }

    waitUntil (condition)

File: KafkaReplication/Distributed.Replication.Kafka.Tests/Utils/TestUtil.cs
using System.Diagnostics;

public static class TestUtil
{
    public static async Task<bool> WaitUntilAsync(int timeoutMs, Func<Task<bool>> asyncCondition, int delay = 100, bool throwException = true)
    {
        return await WaitUntilInternal(timeoutMs, asyncCondition, null, delay, throwException);
    }

    public static Task<bool> WaitUntilAsync(int timeoutMs, Func<bool> condition, int delay = 100, bool throwException = true)
    {
        return WaitUntilInternal(timeoutMs, null, condition, delay, throwException);
    }

    private static async Task<bool> WaitUntilInternal(
        int timeoutMs, 
        Func<Task<bool>>? asyncCondition, 
        Func<bool>? condition, 
        int delay, 
        bool throwException)
    {
        var stopwatch = Stopwatch.StartNew();

        while (true)
        {
            if (stopwatch.ElapsedMilliseconds > timeoutMs)
            {
                if (throwException)
                {
                    throw new TimeoutException("Condition of the test was not met in the specified time");
                }
                return false;
            }

            try
            {
                if (asyncCondition != null && await asyncCondition())
                {
                    return true;
                }
                if (condition != null && condition())
                {
                    return true;
                }
            }
            catch (Exception e)
            {
                // ignored
            }

            await Task.Delay(delay);
        }
    }
}